code/go/0chain.net/chaincore/chain/state/state_context.go

```diff
@@ -400,3 +402,87 @@ func (sc *StateContext) SetStateContext(s *state.State) error {
 func (sc *StateContext) GetLatestFinalizedBlock() *block.Block {
 	return sc.getLatestFinalizedBlock()
 }
+
+type errorIndex struct {
+	err   error
+	index int
+}
+
+type GetItemFunc func(id string, balance CommonStateContextI) (interface{}, error)
+
+type rspIndex struct {
+	index int
+	item  interface{}
+}
+
+// GetItemsByIDs read items by ids from MPT concurrently and safely with consistent values
+// Note: the GetItemFunc should not return custom error that wraps the error returned from
+// StateContextI
+func GetItemsByIDs[T any](ids []string, getItem GetItemFunc, balances CommonStateContextI) ([]T, error) {
+	var (
+		itemC     = make(chan rspIndex, len(ids))
+		stateErrC = make(chan error, len(ids))
+		errC      = make(chan errorIndex, len(ids))
+		wg        sync.WaitGroup
+	)
+
+	for i, id := range ids {
+		wg.Add(1)
+		go func(idx int, id string) {
+			defer wg.Done()
+			item, err := getItem(id, balances)
+			if err != nil {
+				if err != util.ErrValueNotPresent {
+					stateErrC <- err
+					return
+				}
+
+				errC <- errorIndex{
+					err:   err,
+					index: idx,
+				}
+				return
+			}
+
+			itemC <- rspIndex{
+				index: idx,
+				item:  item,
+			}
+		}(i, id)
+	}
+	wg.Wait()
+	close(itemC)
+	close(errC)
+
+	// check internal error first
+	select {
+	case err := <-stateErrC:
+		return nil, err
+	default:
+	}
+
+	errIdxs := make([]errorIndex, 0, len(ids))
+	for ei := range errC {
+		errIdxs = append(errIdxs, ei)
+	}
+
+	if len(errIdxs) > 0 {
+		sort.SliceStable(errIdxs, func(i, j int) bool {
+			return errIdxs[i].index < errIdxs[j].index
+		})
+
+		// we would only return one 'value not present' error (the first one) to avoid too much
+		// error data added to transaction output.
+		logging.Logger.Error("could not get items", zap.Any("errors", errIdxs))
+		retErr := errIdxs[0]
+		return nil, fmt.Errorf("could not get item %q: %v", ids[retErr.index], retErr.err)
+	}
+
+	//ensure original ordering
+	items := make([]T, len(ids))
+	for item := range itemC {
+		items[item.index] = item.item.(T)
```

Comment:
The type cast would fail if the `GetItemFunc` callback function did not implement correctly, adding comma ok would make this kind of error goto transaction as a chargeable error, which is not what we wanted. So we will need to ensure the `getItem` returns the `T` type, I'll do a fix,  then compiler would help us to detect the type errors, which means if we don't implement the `GetItemFunc` with correct type, it wouldn't be able to pass the compiler.

Reply:
the problem is that if there is a bug there and `GetItemFunc()` returns unexpected type, then malicious user can create transaction that will `panic()` on all miners and network will be dead (during block generation). if no panic happens, then this transaction will simply be marked as invalid and nothing strange will happen, even no restart will occur. panics are reasonable only during startups I guess or something really unrecoverable, but any tricky transactions should be processed as usual with chargeable or unchargeable errors

Reply:
The unexpected type would not be able to pass the compiler, it must be the same a the type we specified or nil.